### DISCUSSION TITLE
关于重组作用域有几个问题

### DISCUSSION THREAD

**Discussion by biubiuqiu0** [compose-museum/jetpack-compose-tutorial#129] [Q&A] [answered]
文章里说 "为什么 Button 不参与重组？"，这个不对吧，打断点看内部有重新执行。换一个方式说，内部的Text文案很长，肯定会导致Button重新执行Layout吧（如果没有限定Button宽高或者外部约束啥的）

**Comment by totoneei** [comment]
感谢提出 `Issue`，经过排查，是 `Button` 参数中的 `elevation` 里面的 `interactions` 引起的重组问题
![image](https://user-images.githubusercontent.com/31311826/120661840-e9ada600-c488-11eb-976e-23046645b521.png)

**Comment by 12og3r** [comment, marked as answer]
这个问题是其实是由于Button的点击波纹效果导致的。
在我们创建Button时，Compose会为我们创建一个交互事件源
![image](https://user-images.githubusercontent.com/30253468/120664198-52167a80-c4bd-11eb-99b6-f281edcdd246.png)
这个事件源其中包裹了一个flow实例
![image](https://user-images.githubusercontent.com/30253468/120664281-65294a80-c4bd-11eb-846f-e1f0e4ed2f7f.png)
再回到之前所调用的evelation方法中，evelation仅有这一个默认实现，他在其中声明了一个协程用于监听事件源，当事件源发送数据时，此时在evelation中创建的interactions这个mutableStateList会发生更新，这也导致了整个evelation方法会发生recompose，并且由于外部依赖于evelation这个composable的返回值，所以外部也会参与到recompose中，也就发生了你所看到的整个Button的recompose
![image](https://user-images.githubusercontent.com/30253468/120664923-ea146400-c4bd-11eb-874a-9699006ceed7.png)
这么说你理解了嘛？

**Comment by biubiuqiu0** [comment]
感谢，水波纹动画引起的重组这个明白。还有就是当文本长度变长，Button里面的Surface，Row宽度也会跟着变化，那这个不是重组吗。子view引起父view宽高变化这个是怎么触发的，既然没有触发外层Composable重组，那么是怎么触发重新layout的

**Comment by totoneei** [comment]
`layout` 只是 `button` 内部的一个 **composable** 而已，可能只涉及 `MeasurePolicy` 等一小部分的执行，但是不需要 `Composable`重组。

**Comment by vitaviva** [comment]
大佬对Compose有兴趣的话要不要参与一起研究啊 @Palardin3 ，加我微信13913365856拉你进群啊
